src/cache/cache.go

```diff
@@ -0,0 +1,121 @@
+package cache
+
+import (
+	"os"
+	"sync"
+	"../heap"
+	"../markov"
+)
+
+/********************************
+Cache supports the following external API to users
+
+c.Init(cacheSize int, cacheType CacheType)
+    Initializes a cache with eviction policy and prefetch defined by cache type
+c.Report() (hits, misses)
+    Get a report of the hits and misses  TODO: Do we want a version number or
+    timestamp mechanism of any form here?
+c.Fetch(name string) (*os.File, error)
+
+*********************************/
+type Cache struct {
+	mu          sync.Mutex          // Lock to protect shared access to cache
+	misses		int
+	hits		int
+	cache		map[string]*os.File
+	heap		heap.MinHeap
+	timestamp	int64 // for controlling LRU heap
+	cacheSize	int
+	chain 		*markov.MarkovChain
+	cacheType	CacheType
+}
+
+
+func (c *Cache) Init(cacheSize int, cacheType CacheType) {
+	c.cacheType = cacheType
+	c.misses = 0
+	c.hits = 0
+    c.cacheSize = cacheSize
+	c.cache = make(map[string]*os.File)
+	c.timestamp = 0
+	
+	if cacheType == LRU || cacheType == MarkovEviction {
+		// only LRU caches should use heap
```

Comment:
No, there is no constructor. I can design one if you'd like. The Init() sets struct values.

Reply:
it looks fine I just don't understand how the heap is initialized because c.heap isn't defined before line 44? So I'm just not sure how the heap.Init() knows where to look.

Reply:
the data type of c.heap handles the memory allocation, Init() handles setting default values. I can change this if you'd like

Reply:
addressed in e28ed31